File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/java/exercises/practice/connect/src/main/java/Connect.java
import java.util.LinkedList;
import java.util.Queue;

class Connect {
    private char[][] board;
    private int height;
    private int width;

    public Connect(String[] boardRows) {
        parseBoard(boardRows);
    }

    private void parseBoard(String[] boardRows) {
        height = boardRows.length;
        if (height == 0) {
            board = new char[0][0];
            width = 0;
            return;
        }
        
        board = new char[height][];
        for (int i = 0; i < height; i++) {
            String row = boardRows[i].trim();
            if (row.isEmpty()) {
                board[i] = new char[0];
                continue;
            }
            String[] cells = row.split(" +");
            board[i] = new char[cells.length];
            for (int j = 0; j < cells.length; j++) {
                board[i][j] = cells[j].charAt(0);
            }
        }
        width = (height > 0) ? board[0].length : 0;
    }

    public Winner computeWinner() {
        if (hasOWon()) {
            return Winner.O;
        } else if (hasXWon()) {
            return Winner.X;
        } else {
            return Winner.NONE;
        }
    }

    private boolean hasOWon() {
        if (height == 0 || width == 0) return false;
        boolean[][] visited = new boolean[height][width];
        Queue<int[]> queue = new LinkedList<>();

        // Initialize with all top-row 'O's
        for (int col = 0; col < width; col++) {
            if (board[0][col] == 'O') {
                queue.add(new int[]{0, col});
                visited[0][col] = true;
            }
        }

        int[][] directions = {{-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}};
        while (!queue.isEmpty()) {
            int[] cell = queue.poll();
            int row = cell[0];

            if (row == height - 1) {
                return true;
            }

            for (int[] dir : directions) {
                int newRow = cell[0] + dir[0];
                int newCol = cell[1] + dir[1];

                if (newRow >= 0 && newRow < height && newCol >= 0 && newCol < board[newRow].length
                        && !visited[newRow][newCol] && board[newRow][newCol] == 'O') {
                    visited[newRow][newCol] = true;
                    queue.add(new int[]{newRow, newCol});
                }
            }
        }
        return false;
    }

    private boolean hasXWon() {
        if (height == 0 || width == 0) return false;
        boolean[][] visited = new boolean[height][width];
        Queue<int[]> queue = new LinkedList<>();

        // Initialize with all left-column 'X's
        for (int row = 0; row < height; row++) {
            if (board[row].length > 0 && board[row][0] == 'X') {
                queue.add(new int[]{row, 0});
                visited[row][0] = true;
            }
        }

        int[][] directions = {{-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}};
        while (!queue.isEmpty()) {
            int[] cell = queue.poll();
            int col = cell[1];

            if (col == width - 1) {
                return true;
            }

            for (int[] dir : directions) {
                int newRow = cell[0] + dir[0];
                int newCol = cell[1] + dir[1];

                if (newRow >= 0 && newRow < height && newCol >= 0 && newCol < board[newRow].length
                        && !visited[newRow][newCol] && board[newRow][newCol] == 'X') {
                    visited[newRow][newCol] = true;
                    queue.add(new int[]{newRow, newCol});
                }
            }
        }
        return false;
    }

    public enum Winner {
        O, X, NONE
    }
}

File: src/2025-03-16-12-54-49--a-deepclaude-full-architect-python-0316-2/python/exercises/practice/connect/connect.py

class ConnectGame:
    def __init__(self, board):
        """
        Initialize the game board from a string representation.
        
        Args:
            board: String representation of the game board with '.' for empty cells,
                  'O' for player O's pieces, and 'X' for player X's pieces.
        """
        self.board = []
        for line in board.strip().split('\n'):
            # Remove leading spaces and split by spaces
            stripped = line.lstrip()
            if stripped:
                self.board.append(stripped.split())

    def get_winner(self):
        """
        Determine the winner of the game.
        
        Player O wins by connecting top to bottom.
        Player X wins by connecting left to right.
        
        Returns:
            'O' if player O wins, 'X' if player X wins, or '' if no winner.
        """
        if not self.board:
            return ''
            
        rows = len(self.board)
        
        # Check if player O has connected top to bottom
        from collections import deque
        
        # Find all O's in the top row to start search
        start_cells = []
        for col in range(len(self.board[0])):
            if self.board[0][col] == 'O':
                start_cells.append((0, col))
        
        if start_cells:
            # BFS to find path from top to bottom
            visited = set(start_cells)
            queue = deque(start_cells)
            
            while queue:
                row, col = queue.popleft()
                
                # Check if we've reached the bottom row
                if row == rows - 1:
                    return 'O'  # Player O wins
                
                # Check all six adjacent hexes
                directions = [(-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0)]
                for dr, dc in directions:
                    nr, nc = row + dr, col + dc
                    
                    # Check if the new position is valid
                    if 0 <= nr < rows and 0 <= nc < len(self.board[nr]):
                        if self.board[nr][nc] == 'O' and (nr, nc) not in visited:
                            visited.add((nr, nc))
                            queue.append((nr, nc))
        
        # Check if player X has connected left to right
        # Find all X's in the leftmost column of each row
        start_cells = []
        for row in range(rows):
            if self.board[row][0] == 'X':
                start_cells.append((row, 0))
        
        if start_cells:
            # BFS to find path from left to right
            visited = set(start_cells)
            queue = deque(start_cells)
            
            while queue:
                row, col = queue.popleft()
                
                # Check if we've reached the rightmost column for this row
                if col == len(self.board[row]) - 1:
                    return 'X'  # Player X wins
                
                # Check all six adjacent hexes
                directions = [(-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0)]
                for dr, dc in directions:
                    nr, nc = row + dr, col + dc
                    
                    # Check if the new position is valid
                    if 0 <= nr < rows and 0 <= nc < len(self.board[nr]):
                        if self.board[nr][nc] == 'X' and (nr, nc) not in visited:
                            visited.add((nr, nc))
                            queue.append((nr, nc))
        
        return ''  # No winner yet
